cache/cache_modifier.go

```diff
@@ -0,0 +1,232 @@
+		}
+	}
+
+	return &modifier{
+		db:       db,
+		bucket:   bucket,
+		update:   update,
+		replay:   replay,
+		hermetic: hermetic,
+	}, nil
+}
+
+// modifierFromJSON parses JSON into cache.Modifier.
+//
+// Example JSON Configuration message:
+// {
+//   "file":     "/some/cache.db",
+//   "bucket":   "some_project",
+//   "update":   true,
+//   "replay":   true,
+//   "hermetic": false
+// }
+func modifierFromJSON(b []byte) (*parse.Result, error) {
+	msg := &modifierJSON{}
+	if err := json.Unmarshal(b, msg); err != nil {
+		return nil, fmt.Errorf("cache.Modifier: json.Unmarshal(): %v", err)
+	}
+
+	mod, err := NewModifier(msg.File, msg.Bucket, msg.Update, msg.Replay, msg.Hermetic)
+	if err != nil {
+		return nil, err
+	}
+	return parse.NewResult(mod, msg.Scope)
+}
+
+// Close closes the underlying database.
+func (m *modifier) Close() error {
+	if m.db != nil {
+		return m.db.Close()
+	}
+	return nil
+}
+
+// ModifyRequest modifies the http.Request.
+func (m *modifier) ModifyRequest(req *http.Request) error {
+	if !m.replay {
+		return nil
+	}
+
+	key, err := getCacheKey(req)
+	if err != nil {
+		return fmt.Errorf("cache.Modifier: getCacheKey(): %v", err)
+	}
+
+	if err := m.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(m.bucket))
+		cached := b.Get(key)
+		if cached != nil {
+			res, err := http.ReadResponse(bufio.NewReader(bytes.NewReader(cached)), req)
+			if err != nil {
+				return fmt.Errorf("http.ReadResponse(): %v", err)
+			}
+			ctx := martian.NewContext(req)
+			ctx.SkipRoundTrip()
+			ctx.Set(cachedResponseCtxKey, res)
+			return nil
+		} else if m.hermetic {
+			return fmt.Errorf("in hermetic mode and no cached response found")
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("cache.Modifier: %v", err)
+	}
+	return nil
+}
+
+// ModifyResponse modifies the http.Response.
+func (m *modifier) ModifyResponse(res *http.Response) error {
+	ctx := martian.NewContext(res.Request)
+	cached, ok := ctx.Get(cachedResponseCtxKey)
+	if ok {
+		*res = *cached.(*http.Response)
+	} else if m.update {
+		key, err := getCacheKey(res.Request)
+		if err != nil {
+			return fmt.Errorf("cache.Modifier: getCacheKey(): %v", err)
+		}
+		var buf bytes.Buffer
+		// TODO: Wrap res.Body so res.Write doesn't close it.
+		if err := res.Write(&buf); err != nil {
+			return fmt.Errorf("cache.Modifier: res.Write(): %v", err)
+		}
+		if err := m.db.Update(func(tx *bolt.Tx) error {
+			b := tx.Bucket([]byte(m.bucket))
+			if err := b.Put(key, buf.Bytes()); err != nil {
+				return fmt.Errorf("bucket.Put(): %v", err)
+			}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("cache.Modifier: %v", err)
+		}
+		r, err := http.ReadResponse(bufio.NewReader(&buf), res.Request)
+		if err != nil {
+			return fmt.Errorf("cache.Modifier: http.ReadResponse(): %v", err)
+		}
+		*res = *r
+	}
+	return nil
+}
+
+// getCacheKey gets a cache key from the request context or generates a new one.
+func getCacheKey(req *http.Request) ([]byte, error) {
```

Comment:
What's the benefit of holding the key in the request context? Generating the key on the fly is pretty inexpensive, and this would only be helpful if we were generating it several times over the course of the request / response - and I only see us doing it twice (once on request, once on response).

Reply:
This is so any upstream request modifiers can set a custom cache key, instead of using the naive default one here.  I think storing it in the context is better than in the HTTP header or URL param, so it doesn't pollute the HTTP request.